main_test.go

```diff
@@ -25,3 +31,148 @@ func TestCIRequirements(t *testing.T) {
 		}
 	}
 }
+
+type providerConfig struct {
+	providerType  string
+	username      string
+	password      string
+	url           string
+	defaultBranch string
+}
+
+var providers = []providerConfig{
+	{
+		providerType:  "azdo",
+		username:      "gitops-promotion",
+		password:      os.Getenv("AZDO_PAT"),
+		url:           os.Getenv("AZDO_URL"),
+		defaultBranch: "main",
+	},
+	{
+		providerType:  "github",
+		username:      "gitops-promotion",
+		password:      os.Getenv("GITHUB_TOKEN"),
+		url:           os.Getenv("GITHUB_URL"),
+		defaultBranch: "main",
+	},
+}
+
+func runCommand(command string, args ...string) (string, error) {
+	var cmdline = []string{"ze-binary", command}
+	cmdline = append(cmdline, args...)
+	message, err := run(cmdline)
+	return message, err
+}
+
+func TestProviderE2E(t *testing.T) {
+	for _, p := range providers {
+		t.Run(p.providerType, func(t *testing.T) {
+			if p.url == "" || p.password == "" {
+				t.Skipf("Skipping test since url or password env var is not set")
+			}
+			path := t.TempDir()
+			providerType, err := git.StringToProviderType(p.providerType)
+			require.NoError(t, err)
+
+			testCloneRepository(t, p.url, p.username, p.password, path, p.defaultBranch)
+
+			now := time.Now()
+			tag := now.Format("20060102150405")
+			group := "testgroup"
+			app := "testapp"
+
+			promoteBranchName := fmt.Sprintf("promote/%s-%s", group, app)
+
+			ctx := context.Background()
+
+			// Test DEV
+			newCommandMsgDev, err := runCommand(
+				"new",
+				"-sourcedir", path,
+				"-provider", p.providerType,
+				"-token", p.password,
+				"-group", group,
+				"-app", app,
+				"-tag", tag,
+			)
+			require.NoError(t, err)
+
+			require.Equal(t, "created promotions pull request", newCommandMsgDev)
+
+			path = testCloneRepositoryAndValidateTag(t, p.url, p.username, p.password, p.defaultBranch, group, "dev", app, tag)
+
+			repoDev := testGetRepository(t, path)
+			revDev := testGetRepositoryHeadRevision(t, repoDev)
+
+			testSetStatus(t, ctx, providerType, revDev, group, "dev", p.url, p.password, true)
```

Comment:
For readability, moving this to a new helper `testFirstEnvironment()` would be possible and then having the same helper function for QA and PROD like `testEnvironment()`. Naming is hard and these names are just bad examples.

Reply:
I still think this would be nice, but not a must have.